### PR TITLE
WidgetMap: support generics in objectTypes

### PR DESCRIPTION
**Pull request by fschinkel**
Parse typeArguments while converting an IdeaConstantValue to an ES6Class. In addition, unwrap JSParenthesizedExpression and (for IJ >= 2023.2) TypeScriptExpressionWithTypeArguments. Fix NPE while resolving the export target.

377491

### CODE REVIEW

File: org.eclipse.scout.sdk.core.typescript/src/main/java/org/eclipse/scout/sdk/core/typescript/builder/imports/ES6ImportValidator.java [hunk 1]
```diff
@@ -54,6 +55,9 @@ protected String computeUniqueNameAndRegisterUsage(IDataType type) {
     if (isBuiltInType(name)) {
       return name; // no import required
     }
+    if (type instanceof IES6Class es6Class) {
+      type = es6Class.withoutTypeArguments();
+    }
```
Comment (mvilliger) [inline org.eclipse.scout.sdk.core.typescript/src/main/java/org/eclipse/scout/sdk/core/typescript/builder/imports/ES6ImportValidator.java:60]: Can the typeArguments really be ignored here? Or should they be processed (create imports if required) as well?

File: org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/model/typescript/IdeaConstantValue.kt [hunk 2]
```diff
@@ -38,21 +40,43 @@ open class IdeaConstantValue(val ideaModule: IdeaNodeModule, internal val elemen
...
-        val unwrappedElement = unwrapTypeScriptAsExpression(element)
+        var unwrappedElement = unwrapJSParenthesizedExpression(element)
+        unwrappedElement = unwrapTypeScriptAsExpression(unwrappedElement)
+        unwrappedElement = unwrapTypeScriptExpressionWithTypeArguments(unwrappedElement)
         unwrapJSNewExpression(unwrappedElement)
     }
 
     override fun containingFile(): Optional<Path> = m_containingFile.computeIfAbsentAndGet { Optional.ofNullable(element?.containingFile?.virtualFile?.resolveLocalPath()) }
 
+    protected fun unwrapJSParenthesizedExpression(element: JSElement?) = if (element is JSParenthesizedExpression) element.innerExpression else element
```
Comment (mvilliger) [inline org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/model/typescript/IdeaConstantValue.kt:51]: This only supports one level: (Class<TypeArg>). Nested Parenthesis is not supported. E.g. ((Class<TypeArg>)).
But maybe this is sufficient like this?

**Review by mvilliger — COMMENTED** [review]
Looks good, thank you very much!
I also added one commit with an additional test case for nested complex type arguments.
On merge you can squash this commit into yours if you like.

File: org.eclipse.scout.sdk.core.typescript/src/main/java/org/eclipse/scout/sdk/core/typescript/builder/imports/ES6ImportValidator.java (hunk 1 above)
Comment (fschinkel) [reply org.eclipse.scout.sdk.core.typescript/src/main/java/org/eclipse/scout/sdk/core/typescript/builder/imports/ES6ImportValidator.java:60]: This is just the computation of a unique name for the type. The type arguments are ignored for this computation. So the unique name of a class with type arguments or the same class without type arguments are  the same.
But the problem here is that the type is used as a key in a map of all unique names. Therefore, the types `SmartField` and `SmartField<number>` will result in the names `SmartField` and `SmartField0`. Omitting the type arguments ensures that only the main type is processed, which is the same in this example and it results only in `SmartField`.

But you are right, the type arguments need to be processed as well. This happens in `DataTypeNameEvaluator#nameForSimpleDataType`, which results in unique names and imports for every type argument.

File: org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/model/typescript/IdeaConstantValue.kt (hunk 2 above)
Comment (fschinkel) [reply org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/model/typescript/IdeaConstantValue.kt:51]: I think this is sufficient as using parenthesis is no longer necessary in IJ 2023. Therefore, I think we do not need to spend extra time for something that is very unlikely to occur.

Comment (mvilliger) [reply org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/model/typescript/IdeaConstantValue.kt:51]: I agree

**Pull request merged**